src/entry-client.js

```diff
@@ -32,35 +33,84 @@ export default function initializeClient(createApp, clientOpts) {
         router.beforeResolve((to, from, next) => {
-            router.getMatchedComponents(to)
-                .filter(c => 'vuex' in c && !registeredModules[c.vuex.moduleName])
-                .forEach((c) => {
-                    opts.logger.info('Registering dynamic Vuex module:', c.vuex.moduleName);
-                    store.registerModule(c.vuex.moduleName, c.vuex.module, {
-                        preserveState: store.state[c.vuex.moduleName] != null,
+            try {
+                router.getMatchedComponents(to)
+                    .filter(c => 'vuex' in c)
+                    .forEach((c) => {
+                        const moduleName = getModuleName(c, to);
+
+                        // Short circuit if we already have this module
+                        const idx = registeredModules.indexOf(moduleName);
+                        if (idx >= 0) {
+                            // Bump this module back to the end of the queue
+                            registeredModules.splice(idx, 1);
+                            registeredModules.push(moduleName);
+                            opts.logger.info(
+                                'Skipping duplicate Vuex module registration:',
+                                moduleName,
+                            );
+                            return;
+                        }
+
+                        opts.logger.info('Registering dynamic Vuex module:', moduleName);
+                        store.registerModule(moduleName, c.vuex.module, {
+                            preserveState: store.state[moduleName] != null,
+                        });
+                        registeredModules.push(moduleName);
                     });
-                    registeredModules[c.vuex.moduleName] = true;
-                });
-            next();
+            } catch (e) {
+                opts.logger.error('Caught error during beforeResolve', e);
+            } finally {
+                next();
+            }
         });
 
         // After routing, unregister any dynamic Vuex modules from prior components
         router.afterEach((to, from) => {
-            const components = router.getMatchedComponents(to);
             const priorComponents = router.getMatchedComponents(from);
 
             priorComponents
-                .filter(c => !includes(components, c) &&
-                             'vuex' in c &&
-                             registeredModules[c.vuex.moduleName])
+                .filter(c => 'vuex' in c)
                 .forEach((c) => {
-                    opts.logger.info('Unregistering dynamic Vuex module:', c.vuex.moduleName);
-                    store.unregisterModule(c.vuex.moduleName);
-                    registeredModules[c.vuex.moduleName] = false;
+                    const toModuleName = getModuleName(c, to);
+                    const fromModuleName = getModuleName(c, to);
+
+                    // After every routing operation, perform available cleanup
```

Comment:
I'm still not sure this is as clean as it should be, need to go back in and look at this with a clearer head